### PR TITLE
Add new OctaveConvolution Layer to keras

### LINKED ISSUE #37576: Add Octave Convolution Layers to tensorflow keras layers
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): TF 2.1
- Are you willing to contribute it (Yes/No): Yes. And the code change is ready.

**Describe the feature and the current behavior/state.**
_Feature:_
[Drop an Octave: Reducing Spatial Redundancy in Convolutional Neural Networks with Octave Convolution](http://openaccess.thecvf.com/content_ICCV_2019/papers/Chen_Drop_an_Octave_Reducing_Spatial_Redundancy_in_Convolutional_Neural_Networks_ICCV_2019_paper.pdf)

_Main idea:_
Decompose the feature maps of the convolutional layers into features of different spatial frequencies which will enlarge the receptive field and reduce the FLOPs and memory cost.

Add an implementation of octave convolution layers (OctConv1D, OctConv2D, OctConv3D) and octave transposed convolution layers (OctConv2DTranspose, OctConv3DTranspose) that inherit from an abstract OctConv layer which, in turn, inherits from tf.keras.layers.Layer (same hierarchy as the already existing convolutional layers).
An octave_add function that takes as input the output of an OctConvND layer (list of two tensors one for  high frequencies and the other for low frequencies) and a tf.keras.layer like Dropout, MaxpoolingND, etc which will be applied on the inputs. The output is the list of the two resulting tensors

The backbone of the implementation is based on the following [github repository](https://github.com/CyberZHG/keras-octave-conv) (MIT and "Anti 996" licensed) by @CyberZHG

The original code has been substantially modified:
 - migration from keras to tf.keras
 - implementation of an abstract OctaveConv layer to emulate the hierarchy of the already existing Conv layers in TensorFlow
 - implementation of the 3D octave convolution layer OctConv3D and the transposed octave convolution layers OctConv2DTranspose, OctConv3DTranspose
 - bug fixes (for instance, in the original implementation setting the trainable attribute of the octave convolution layers to False wouldn't freeze the learnable weights)

_Sample use case of the layers OctConvLayers:_

```
l1 = Input(shape=(28,28, 1))
l2 = OctaveConv2D(32, (3, 3), activation='relu', low_freq_ratio=0.25)(l1_input) # takes a single input tensor and outputs a list of 2 output tensors (low_freq_ratio > 0)
l3 = octave_add(l2, MaxPooling2D(pool_size=(2, 2)))
l4 = octave_add(l3, Dropout(0.2))
l5 = OctaveConv2D(32, (3, 3), activation='relu', low_freq_ratio=0)(l4) # takes a list of 2 tensors and outputs a tensor (low_freq_ratio = 0)
l6 = Flatten()(l5) # l5 is a tensor so no need to use octave_add
```

_Current Behavior:_
Currently tensorflow doesn't have any implementation of the octave convolution layers.

_Why we need this api:_
The "vanilla" ConvND/ConvNDTranpose layers can be directly replaced by the OctConvND/OctConvNDTranspose layers and simple experiments that will be provided show that models using octave convolutions consistently give better performances than their vanilla versions.
Also, using octave convolutions will reduce the FLOPs and the memory cost.

**Will this change the current api? How?**
No, it won't change the current api.
It will add 6 new classes to tf.keras.layers (1 abstract, 3 OctConvND and 2 OctConvNDTranspose)

**Who will benefit with this feature?**
TensorFlow users who are seeking to improve the performances and computational costs of their model with a direct replacement of the vanilla convolutions without any adjustments in the network architecture.

**Any Other info.**

### PR DESCRIPTION
**Pull request by maxkaustav**
In natural images, information is conveyed at different frequencies where higher frequencies are usually encoded with fine details and lower frequencies are usually encoded with global structures. Similarly, the output feature maps of a convolution layer can also be seen as a mixture of information at different frequencies. In this work, we propose to factorize the mixed feature maps by their frequencies and design a novel Octave Convolution (OctConv) opera-tion1to store and process feature maps that vary spatially“slower” at a lower spatial resolution reducing both memory and computation cost.
Fixes #37576
https://arxiv.org/pdf/1904.05049.pdf
Here is a test code on mnist dataset

[mnistoctave.zip](https://github.com/tensorflow/tensorflow/files/4404609/mnistoctave.zip)

### CODE REVIEW

**Comment by AchilleMascia** [comment]
Hi,
I'm the author of the issue (#37576 ) mentioned in this PR.
This PR states that it fixes it but it doesn't for the reasons I give in my PR (#38154) for this issue and also in a comment in the issue.
I therefore invite you and the tensorflow team to look at the code I propose in #38154 instead.

**Comment by gabrieldemarmiesse** [comment]
I believe this pull request should be closed and that we should group our efforts to work on the pull request in thensorflow addons instead.
